### ISSUE TITLE
Future.firstCompletedOf introduces a memory leak when at least one of the futures is long-lived

### ISSUE THREAD

**Issue by eyalfa** [scala/bug#10513]
When using Future.firstCompletedOf with at least one long-lived future (long lived = either unlikely to ever complete, or may take relatively long time to complete) the program accumulates memory until all futures passed to FuturefirstCompledOf are actually complete.

I've created a small [repo](https://github.com/eyalfa/future-firstCompleted-test/blob/master/src/test/scala/scala/concurrent/FutureFirstCompletedSuite.scala#L18) with a failing test case and a partial fix.

root cause:
Future.firstCompletedOf is implemented in terms of `onComplete` registering a handler that tries to complete the result promise, the first completed future actually gets to complete the result promise with its own result, now the rest of the future has a references path from their list of handlers to the registered handler to the completed promise to the first result. this references path is broken only once these futures complete, if at leas one of the futures is long-lived this chain won't be GCed for a long/indefinite period of time.

my proposed partial fix replaces the handler function with a bit more elaborate handler that clears the reference for the resulting promise after completing it, this breaks the reference chain but leaves the handlers registered on all other futures, I think a more complete approach would attempt to remove these handlers.

CC: @axel22

**Comment by axel22** [comment]
I see - basically, this problem occurs if the value in the resulting future has a large footprint. Your solution looks legitimate to me! Perhaps you could consider submitting a patch?

**Comment by eyalfa** [comment]
Most definitely, will attend to this later this week.

**Comment by eyalfa** [comment]
@axel22 what about the potentially ever-growing list of handlers? do you want to address this? perhaps in a separate PR?

**Comment by axel22** [comment]
Thanks! I'm not completely sure what you mean by the ever-growing list of handlers, but it sounds like it could be addressed separately, in a different PR.

**Comment by eyalfa** [comment]
in my original use case the 'long-lived' promise was actually there to mark an abort-flag.
so this promise was rarely ever completed, but every future created in the system was 'merged' with it using the `firstCompletedOf` combinator, so after some (relatively short) time this promise held references to almost every historic response and intermediate result in the system.

**Comment by axel22** [comment]
I see - thanks for explaining. That maybe sounds like a separate problem to me, but I'd say it's up to you if you want to separate the solution or keep it in the same PR.

**Comment by eyalfa** [comment]
I say, let's get done with what we're confident with and then try moving on to the next step

**Comment by axel22** [comment]
SGTM!

**Comment by viktorklang** [comment]
@eyalfa PRs are most welcome :)

**Comment by eyalfa** [comment]
@viktorklang , I'll submit a PR (just try to stop me! :sunglasses:),
I hope I'll find the time in the following few days (we go on holidays here)

**Comment by viktorklang** [comment]
@eyalfa Awesome, looking forward to having a look at it! :)

**Comment by jchapuis** [comment]
👋 i realize this thread is old but we have experienced a severe memory leak directly linked to our usage of `firstCompletedOf`. The symptom was a very high count of `Promise$Transformation` instances. We were invoking the function with a short-lived and a long-lived Future, the latter representing the termination of some resource.

There is something really puzzling in the current implementation (see below): we register the `firstCompleteHandler` callback for all futures with ` i.next().onComplete(firstCompleteHandler)`, but we never "unregister" callbacks to other futures. Since in our case the short-lived Future almost always completes first, the very same long-lived Future instance accumulates a large number of callbacks which increases the instance count to large levels. 

Any thoughts?
   
```scala
final def firstCompletedOf[T](futures: IterableOnce[Future[T]])(implicit executor: ExecutionContext): Future[T] = {
    val i = futures.iterator
    if (!i.hasNext) Future.never
    else {
      val p = Promise[T]()
      val firstCompleteHandler = new AtomicReference[Promise[T]](p) with (Try[T] => Unit) {
        override final def apply(v1: Try[T]): Unit =  {
          val r = getAndSet(null)
          if (r ne null)
            r tryComplete v1 // tryComplete is likely to be cheaper than complete
        }
      }
      while(i.hasNext && firstCompleteHandler.get != null) // exit early if possible
        i.next().onComplete(firstCompleteHandler)   
      p.future
    }
  }
```

**Comment by SethTisue** [comment]
@jchapuis what Scala version?

**Comment by som-snytt** [comment]
The OP comment ends, "...but leaves the handlers registered on all other futures, I think a more complete approach would attempt to remove these handlers." So apparently it was considered. Is that due consideration?

**Comment by adriaanm** [comment]
We're on 2.13.10 (also, 👋 )

**Comment by axel22** [comment]
It seems that the only way to solve this is to have an API on futures to deregister callbacks, which `firstCompletedOf` could then use.

**Comment by som-snytt** [comment]
> to have an API on futures

or an internal API. I just joked somewhere "they're all good dogs", which in this context means, "They're all DefaultPromises", if they want this optimization.

**Comment by jchapuis** [comment]
> It seems that the only way to solve this is to have an API on futures to deregister callbacks, which `firstCompletedOf` could then use.

yes, that would seems like the most API-preserving way to proceed

another (a bit convoluted) option I can see would be to introduce some stateful construct specialized to this use case of racing short-lived futures with a long-running one. One could create such a "race manager" for a specific future and tie its lifetime to the Future instance. Something like `Future.createRaceManager`. The returned stateful wrapper class could offer a small API such as `raceWith(futures: IterableOnce[Future[T]]): Future[T]`.  Upon a call to `raceWith`, it creates (and returns) a promise. It subscribes a callback to each future in `futures` to complete this promise. It also registers this promise into a concurrent set of all outstanding promises, so that the long-running future (the instance member) can complete them upon its own completion. Upon completion of the returned promise, it's also removed from the set, so that we don't leak. 
This solution doesn't solve the issue that we register callbacks for all the ones we pass in `futures` though. But the documentation could state that these are assumed short-lived.

Not ideal though, would rather be in favor of what was mentioned above 🙂

**Comment by nmarton-da** [comment]
Just came across this, is this already resolved?
Should this issue reopened/another one created?
It would be so great to solve this, as it is so easy to fall into this trap in the one long-lived/huge amount of short-lived scenario.

**Comment by lrytz** [comment]
Let's put this on the radar, can someone create a new ticket?

**Comment by som-snytt** [comment]
I created a placeholder ticket.
